ctms/bin/acoustic.py

```diff
@@ -255,19 +257,26 @@ def do_dump(dbsession, contacts, output: TextIO):
         m.source: m.destination for m in get_all_acoustic_newsletters_mapping(dbsession)
     }
 
-    fieldnames = None
-    writer = None
-    for email in contacts:
-        contact = ContactSchema.from_email(email)
-        main_table_row, _, _ = service.convert_ctms_to_acoustic(
-            contact, main_fields, newsletters_mapping
-        )
-        # Write header on the first iteration.
-        if fieldnames is None:
-            fieldnames = sorted(main_table_row.keys())
-            writer = csv.DictWriter(output, fieldnames=fieldnames)
-            writer.writeheader()
-        writer.writerow(main_table_row)
+    # Do a first pass to obtain an exhaustive list of columns.
+    # We use a temporary file to avoid loading everything in memory.
+    columns = set()
+    with tempfile.NamedTemporaryFile(mode="w") as tmpfile:
+        for email in contacts:
+            contact = ContactSchema.from_email(email)
+            main_table_row, _, _ = service.convert_ctms_to_acoustic(
+                contact, main_fields, newsletters_mapping
+            )
+            columns |= set(main_table_row.keys())
```

Comment:
Slick, but I'm always in favor of the explicit names for set operations 😅 
```suggestion
            columns.update(main_table_row.keys())
```